src/java/org/apache/cassandra/schema/IndexMetadata.java

```diff
@@ -103,30 +111,143 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
+    {
+        String indexNameUntruncated = PATTERN_NON_WORD_CHAR.matcher(table + '_' + column.toString()).replaceAll("");
+        String indexNameTrimmed = indexNameUntruncated
+                                  .substring(0,
+                                             Math.min(calculateGeneratedIndexNameLength(keyspaceNameLength, table.length()),
+                                                      indexNameUntruncated.length()));
+        return indexNameTrimmed + INDEX_POSTFIX;
+    }
+
+    /**
+     * Calculates the allowed length of the index name to fit file names,
+     * so it is guaranteed to fit the file name length specified by name length constant.
+     *
+     * @param keyspaceNameLength the length of the keyspace name
+     * @param tableNameLength    the length of the table name
+     * @return the allowed length of the index name
+     */
+    private static int calculateIndexAllowedLength(int keyspaceNameLength, int tableNameLength)
+    {
+
+        int addedLength1 = getAddedLengthFromIndexContextFullName(keyspaceNameLength, tableNameLength);
+        int addedLength2 = getAddedLengthFromDescriptorAndVersion();
+
+        int maxAddedLength = Math.max(addedLength1, addedLength2);
+
+        assert maxAddedLength <= SchemaConstants.NAME_LENGTH : "Index name additions are too long";
+
+        return SchemaConstants.NAME_LENGTH - maxAddedLength;
+    }
+
+    /**
+     * Calculates the maximum length of the generated index name to fit file names.
+     * It includes the generated suffixes in account.
+     *
+     * @param keyspaceNameLength the length of the keyspace name
+     * @param tableNameLength    the length of the table name
+     * @return the allowed length of the generated index name
+     */
+    private static int calculateGeneratedIndexNameLength(int keyspaceNameLength, int tableNameLength)
+    {
+        int uniquenessSuffixLength = 3;
+        int indexNameAddition = uniquenessSuffixLength + INDEX_POSTFIX.length();
+        int allowedIndexNameLength = calculateIndexAllowedLength(keyspaceNameLength, tableNameLength);
+        if (allowedIndexNameLength < indexNameAddition)
+            throw new InvalidQueryException(String.format("Cannot generate index name to fit file names, since the addition take the allowed length %s.", SchemaConstants.NAME_LENGTH));
+        return allowedIndexNameLength - indexNameAddition;
+    }
+
+    /**
+     * Calculates the length of the added prefixes and suffixes from Descriptor constructor
+     * and Version.stargazerFileNameFormat.
+     *
+     * @return the length of the added prefixes and suffixes
+     */
+    private static int getAddedLengthFromDescriptorAndVersion()
+    {
+        int separatorLength = 1;
+        // Prefixes and suffixes constructed by Version.stargazerFileNameFormat
+        int versionNameLength = Version.latest().toString().length();
+        int generationLength = 1;
+        int addedLength = SAI_DESCRIPTOR.length()
```

Comment:
This method is based on SAI-specific concepts, whereas `IndexMetadata` should be implementation-agnostic. Not sure what will happen when generating index names for the other available index implementations. I think we should move this to the `Index` interface and implement it in `SotrageAttachedIndex`.

Reply:
Moved into SAI